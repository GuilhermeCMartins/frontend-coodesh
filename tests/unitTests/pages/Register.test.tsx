import { act } from "react-dom/test-utils";
import useAuthentication from "../../../src/app/hooks/useAutentication";
import Register from "../../../src/app/register/page";
import {
  RenderResult,
  cleanup,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import { ToastContainer } from "react-toastify";
import { ReactElement } from "react";

const renderWithToastify = (component: ReactElement): RenderResult => {
  return render(
    <div>
      <ToastContainer />
      {component}
    </div>
  );
};

jest.mock("next/navigation", () => ({
  useRouter: () => ({
    push: jest.fn(),
  }),
}));

jest.mock("../../../src/app/hooks/useAutentication", () => ({
  __esModule: true,
  default: jest.fn(() => ({
    register: jest.fn().mockImplementation((name, password, type) => {
      if (name == "John") {
        throw Error;
      }
      return Promise.resolve({
        success: true,
      });
    }),
  })),
}));

describe("Register", () => {
  afterEach(() => {
    cleanup();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<Register />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("registers successfully with valid name, password", async () => {
    renderWithToastify(<Register />);

    const registerButton = screen.getByRole("button", { name: "Register" });

    const nameInput = screen.getByLabelText("Nome");
    const passwordInput = screen.getByLabelText("Senha");

    act(() => {
      fireEvent.change(nameInput, { target: { value: "JohnDoe" } });
      fireEvent.change(passwordInput, { target: { value: "password123" } });
    });

    act(() => {
      fireEvent.click(registerButton);
    });

    expect(
      await screen.findByText("Registrado com sucesso!")
    ).toBeInTheDocument();
  });

  it("displays error messages for missing name and password", async () => {
    render(<Register />);

    const registerButton = screen.getByRole("button", { name: "Register" });

    fireEvent.click(registerButton);

    await waitFor(() =>
      expect(useAuthentication().register).not.toHaveBeenCalled()
    );

    const nameErrorMessages = screen.getAllByTestId("name-error");
    const passwordErrorMessages = screen.getAllByTestId("password-error");

    expect(nameErrorMessages[0]).toBeInTheDocument();
    expect(passwordErrorMessages[0]).toBeInTheDocument();
  });

  it("displays error message for existing user", async () => {
    renderWithToastify(<Register />);

    const registerButton = screen.getByRole("button", { name: "Register" });

    const nameInput = screen.getByLabelText("Nome");
    const passwordInput = screen.getByLabelText("Senha");

    act(() => {
      fireEvent.change(nameInput, { target: { value: "John" } });
      fireEvent.change(passwordInput, { target: { value: "password123" } });
    });

    act(() => {
      fireEvent.click(registerButton);
    });

    await waitFor(() =>
      expect(screen.getByText("Usuário já existente.")).toBeInTheDocument()
    );
  });
});
