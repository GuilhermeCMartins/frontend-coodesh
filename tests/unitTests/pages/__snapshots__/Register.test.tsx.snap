// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="c-csXaew"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm css-cuefkz-MuiContainer-root"
      style="display: flex; flex-direction: column; justify-content: center; align-items: center; box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1); padding: 2rem; border-radius: 8px; background-color: white; min-height: 40vh; border-top: 4px solid #2196f3;"
    >
      <h6
        class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
        style="margin-bottom: 1rem; text-transform: uppercase; font-weight: 700;"
      >
        Register
      </h6>
      <div
        class="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          data-testid="name-error"
          style="margin-bottom: 1rem;"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r0:"
            id=":r0:-label"
          >
            Nome
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              placeholder="Digite seu nome"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Nome
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          data-testid="password-error"
          style="margin-bottom: 1rem;"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r1:"
            id=":r1:-label"
          >
            Senha
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r1:"
              placeholder="Digite sua senha"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Senha
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
          id="typeVendor"
        >
          Escolha o tipo de vendedor
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth  css-1ualgfl-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
          style="margin-bottom: 1rem;"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="typeVendor typeVendor"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="typeVendor"
            role="button"
            tabindex="0"
          >
            Produtor
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            tabindex="-1"
            value="Creator"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Vendedor
              </span>
            </legend>
          </fieldset>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Register
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        style="margin-top: 1rem;"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          Já possui uma conta? 
        </span>
        <a
          href="/"
          style="text-decoration: none; color: gray;"
        >
          Faça o login
        </a>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
